istioctl/pkg/multixds/gather.go

```diff
@@ -113,6 +120,110 @@ func queryEachShard(all bool, dr *xdsapi.DiscoveryRequest, istioNamespace string
 	return responses, nil
 }
 
+// queryDebugSynczViaAgents sends a debug/syncz xDS request via Istio Agents.
+// By this way, even if istioctl cannot access a specific `istiod` instance directly,
+// `istioctl` can access the debug endpoint.
+// If `all` is true, `queryDebugSynczViaAgents` iterates all the pod having a proxy
+// except the pods of which status information is already queried.
+func queryDebugSynczViaAgents(all bool, dr *xdsapi.DiscoveryRequest, istioNamespace string, kubeClient kube.CLIClient,
+	centralOpts clioptions.CentralControlPlaneOptions,
+) ([]*xdsapi.DiscoveryResponse, error) {
+	xdsOpts := clioptions.CentralControlPlaneOptions{
+		XDSSAN:  makeSan(istioNamespace, kubeClient.Revision()),
+		CertDir: centralOpts.CertDir,
+		Timeout: centralOpts.Timeout,
+	}
+	visited := make(map[string]bool)
+	queryToOnePod := func(pod *v1.Pod) (*xdsapi.DiscoveryResponse, error) {
+		fw, err := kubeClient.NewPortForwarder(pod.Name, pod.Namespace, "localhost", 0, 15004)
+		if err != nil {
+			return nil, err
+		}
+		err = fw.Start()
+		if err != nil {
+			return nil, err
+		}
+		defer fw.Close()
+		xdsOpts.Xds = fw.Address()
+		// Use plaintext.
+		response, err := xds.GetXdsResponse(dr, istioNamespace, tokenServiceAccount, xdsOpts, []grpc.DialOption{})
+		if err != nil {
+			return nil, fmt.Errorf("could not get XDS from the agent pod %q: %v", pod.Name, err)
+		}
+		for _, resource := range response.GetResources() {
+			switch resource.GetTypeUrl() {
+			case "type.googleapis.com/envoy.service.status.v3.ClientConfig":
+				clientConfig := xdsstatus.ClientConfig{}
+				err := resource.UnmarshalTo(&clientConfig)
+				if err != nil {
+					return nil, err
+				}
+				visited[clientConfig.Node.Id] = true
+			default:
+				// ignore unknown types.
+			}
+		}
+		return response, nil
+	}
+
+	responses := []*xdsapi.DiscoveryResponse{}
+	if all {
+		token := ""
+		touchedPods := 0
+
+	GetProxyLoop:
+		for {
+			list, err := kubeClient.GetProxyPods(context.TODO(), int64(kubeClientGetPodLimit), token)
+			if err != nil {
+				return nil, err
+			}
+			// Iterate all the pod.
+			for _, pod := range list.Items {
+				touchedPods++
+				if centralOpts.XdsViaAgentsLimit != 0 && touchedPods > centralOpts.XdsViaAgentsLimit {
+					break GetProxyLoop
```

Comment:
In this case we should have output alerting the user that some data may be missing.  It's good to have an upper limit here, we just need to make sure the user knows they've hit the limit, and might not have complete data.

Reply:
Totally agree with it. What could be proper way to display the message? stderr?

Reply:
Hm, I refactored it a little bit, and displayed the message. Please take a look.